contracts/message/apps/RFQ.sol

```diff
@@ -0,0 +1,244 @@
+    mapping(uint64 => uint32) public feePercOverride;
+    uint64 public safeTime;
+
+    event Msg1Sent(bytes32 hash, AgreementDetail detail, address srcRecipient, uint64 submissionDeadline);
+    event Msg2Sent(bytes32 hash);
+    event Msg3Sent(bytes32 hash);
+    event MessageReceived(bytes32 hash);
+    event Msg2Executed(bytes32 hash, address srcRecipient, address srcToken, uint256 amount);
+    event Msg3Executed(bytes32 hash, address refundTo, address srcToken, uint256 amount);
+    event AppUpdated(uint64 chainId, address app);
+    event FeePercUpdated(uint64 chainId, uint32 feePerc);
+    event SafeTimeUpdated(uint64 safeTime);
+
+    constructor(address _messageBus) {
+        messageBus = _messageBus;
+    }
+
+    function sendMessage1(AgreementDetail calldata _agrDetail, uint64 _submissionDeadline)
+        external
+        payable
+        whenNotPaused
+        returns (bytes32)
+    {
+        require(_submissionDeadline > block.timestamp, "past submission deadline");
+        require(_agrDetail.releaseDeadline > (block.timestamp + safeTime), "inappropriate release deadline");
+        require(
+            _agrDetail.dstRecipient != address(0) &&
+                _agrDetail.refundTo != address(0) &&
+                _agrDetail.srcRecipient != address(0),
+            "src/dstRecipient, refundTo should not be 0 address"
+        );
+        require(_agrDetail.srcChainId == uint64(block.chainid), "mismatch src chainId");
+        require(_agrDetail.usrAddr == msg.sender, "mismatch usr addr");
+
+        bytes32 agrHash = getAgrHash(_agrDetail);
+        require(orderBooks[agrHash] == false, "still pending order");
+        uint256 rfqFee = getRFQFee(_agrDetail.dstChainId, _agrDetail.amount1);
+        require(rfqFee <= _agrDetail.amount1, "too small amount to cover protocol fee");
+        IERC20(_agrDetail.srcToken).safeTransferFrom(msg.sender, address(this), _agrDetail.amount1);
+        orderBooks[agrHash] = true;
+
+        address _receiver = apps[_agrDetail.dstChainId];
+        require(_receiver != address(0), "no rfq contract on dst chain");
+        bytes memory message = abi.encode(agrHash);
+        sendMessage(_receiver, _agrDetail.dstChainId, message, msg.value);
+        emit Msg1Sent(agrHash, _agrDetail, _agrDetail.srcRecipient, _submissionDeadline);
+        return agrHash;
+    }
+
+    function sendMessage2(AgreementDetail calldata _agrDetail) external payable whenNotPaused {
+        require(_agrDetail.releaseDeadline > block.timestamp, "past release deadline");
+        require(_agrDetail.dstChainId == uint64(block.chainid), "mismatch dst chainId");
+        bytes32 agrHash = getAgrHash(_agrDetail);
+        require(filledOrder[agrHash] == false, "order already filled");
+        IERC20(_agrDetail.dstToken).safeTransferFrom(msg.sender, _agrDetail.dstRecipient, _agrDetail.amount2);
+        filledOrder[agrHash] = true;
+
+        address _receiver = apps[_agrDetail.srcChainId];
```

Comment:
can't we just pass in the receiver address along with AgreementDetail? saving every app addr on every chain sounds like lots of ops overhead

Reply:
> can't we just pass in the receiver address alongside AgreementDetail? saving every app addr on every chain sounds like lots of ops overhead

We have to have these addrs configured on chain as they are required for verification when executing messages.